com.ibm.wala.core/src/com/ibm/wala/ipa/slicer/SDGSupergraphLightweight.java

```diff
@@ -0,0 +1,447 @@
+package com.ibm.wala.ipa.slicer;
+
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.google.common.collect.Table;
+import com.google.common.collect.Table.Cell;
+import com.ibm.wala.dataflow.IFDS.ISDGSupergraph;
+import com.ibm.wala.ipa.slicer.Statement.Kind;
+import com.ibm.wala.util.collections.EmptyIterator;
+import com.ibm.wala.util.debug.Assertions;
+import com.ibm.wala.util.graph.Graph;
+import com.ibm.wala.util.intset.IntSet;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Class providing lightweight representation of an SDG for slicing. Every
+ * Statement is represented by a Long, where the bits are allocated as follows,
+ * indexing from the right:
+ * 
+ * bits 35-63 - 29 bits to encode the pdgId (id of call graph node and PDG)
+ * 
+ * bits 5-34 - 30 bits to encode the localId (id of the Statement within the
+ * respective PDG)
+ * 
+ * bits 1-4 - 4 bits encodes the Statement.Kind (16 options)
+ * 
+ * bit 0 - result of SDGSupergraph.isCall (depends on control dependence
+ * settings so must be stored separately)
+ */
+@JsonDeserialize(using = com.ibm.wala.ipa.slicer.json.SDGSupergraphLightweightDeserializer.class)
+public class SDGSupergraphLightweight implements ISDGSupergraph<Long, Integer> {
+
+  // graph edge info
+  // "b succ of a" does not necessarily imply "a pred of b" or vice versa
+  // this lack of implication is also true in standard WALA SDGs
+  // e.g. if code being analyzed uses Class.newInstance()
+  private final Map<Long, Set<Long>> successors;
+  private final Map<Long, Set<Long>> predecessors;
+
+  // map every procedure (i.e., pdgId) to the set of entry/exit nodes
+  // see SDGSupergraph.getEntriesForProcedure() and getExitsForProcedure()
+  private final Map<Integer, Long[]> procEntries;
+  private final Map<Integer, Long[]> procExits;
+
+  // maps statement to call site if stmt has one
+  private final Map<Long, Integer> stmtsToCallSites;
+
+  // pdgId -> call site -> caller param and return statements
+  private final Table<Integer, Integer, Set<Long>> callStmtsForSite;
+  private final Table<Integer, Integer, Set<Long>> retStmtsForSite;
+
+  // maps used to compare locations for HeapStatements
```

Comment:
I agree it's a bit cleaner even though it may not improve performance a lot (it definitely doesn't show up in profiling as a bottleneck). Will change.